### DISCUSSION TITLE
Socket.io using Redis adapter complained on server.handleUpgrade() called more than once

### DISCUSSION THREAD

**Discussion by yturgema1** [socketio/socket.io#4584] [Q&A] [answered]
Hey 
we're using in Node.js on server and React on client on Azure. [   "@socket.io/redis-adapter": "^8.0.0",  "socket.io": "^4.4.1",    "redis": "^4.1.0"], [ "socket.io-client": "^4.4.1"]

**server code:**
**socket.js** -
```
let io;
module.exports = {
	init: (httpServer) => {
		io = require('socket.io')(httpServer, {
			perMessageDeflate: false,
			cors: {
				origin: process.env.CLIENT_HOST_NAME,
				methods: ['GET', 'POST', 'DELETE', ' PUT', 'OPTIONS'],
			},
		});
		return io;
	},

	getIO: () => {
		if (!io) {
			throw new Error('Socket.io not initialized!');
		}

		return io;
	},
};
```

**index.js -**

```
const express = require('express');
const http = require('http');
const app = express();
const httpServer = http.createServer(app);
const createAdapter = require('@socket.io/redis-adapter').createAdapter;
const redis = require('redis');
const { createClient } = redis;

httpServer.server = httpServer.listen(port, async () => {
	console.log(`Running on port: ${port}`);
});
const io = require('./shared/socket').init(httpServer);
app.use(express.static(path.join(__dirname, 'public')));

(async () => {
	try {
		const cacheHostName = process.env.REDISCACHEHOSTNAME;
		const cachePassword = process.env.REDISCACHEKEY;
		let pubClient = createClient({
			url: `rediss://${cacheHostName}`,
			password: cachePassword,
		});
		await pubClient.connect();
		let subClient = pubClient.duplicate();
		io.attach(httpsServer);
		io.adapter(createAdapter(pubClient, subClient));
	} catch (error) {
		console.error(`SOCKET - ${error}`);
	}
})();

io.on('connection', (socket) => {
	console.log('Debug IO Adapter = ', io.of('/').adapter);
	socket.on('disconnect', (reason) => {
		console.log(`DISCONNECT REASON - ${reason}`);
	});
	socket.on('error', (error) => {
		console.error(`IO ERROR - ${JSON.stringify(error)}`);
	});
});
```

**Client code:**

```
import { io } from 'socket.io-client';

const socketConnection = {
	_instance: null,
	getInstance() {
		if (!this._instance) {
				console.log('Azure socket instance ');
				return io(process.env.REACT_APP_AKS_BACKEND, {
					path: '/ffcbackend/socket.io/',
					transports: ['websocket'],
					forceNew: false,
					// forceNew: true,
				});
		}
		return this._instance;
	},
};
```

this code working perfect on local, when we tried to deploy on azure we got an error:

_Error: server.handleUpgrade() was called more than once with the same socket, possibly due to a misconfiguration
    at WebSocketServer.completeUpgrade (/usr/src/node_modules/engine.io/node_modules/ws/lib/websocket-server.js:340:13)
    at WebSocketServer.handleUpgrade (/usr/src/node_modules/engine.io/node_modules/ws/lib/websocket-server.js:316:10)
    at /usr/src/node_modules/engine.io/build/server.js:397:21
    at Server.verify (/usr/src/node_modules/engine.io/build/server.js:127:24)
    at Server.handleUpgrade (/usr/src/node_modules/engine.io/build/server.js:383:14)
    at Server.<anonymous> (/usr/src/node_modules/engine.io/build/server.js:495:26)
    at Server.emit (node:events:525:35)
    at onParserExecuteCommon (node:_http_server:772:14)
    at onParserExecute (node:_http_server:686:3)
    at HTTPParser.callbackTrampoline (node:internal/async_hooks:130:17)_

any idea why ?
what we're missing ?

**Comment by darrachequesne** [comment, marked as answer]
Hi! I think that's because the HTTP server is attached twice:

Here:

```js
const io = require('./shared/socket').init(httpServer);
```

And here:

```js
io.attach(httpsServer); // I guess it's a typo and should be "httpServer" instead
```

Could you please check?

**Reply by yturgema1** [reply]
yes it's worked but only after except for your suggestion we changed the following:

1. moved the httpsServer.listen(....) to be below the io.on(...) 
2. run pubClient.duplicate() before the connect()
3. run the connect for the pub&sub in Promise.all -> and only then the io.adapter(createAdapter(...))

In conclusion it's doesn't work with await... from some reason

Thanks a lot dude
